File: takengine/takengine/src/main/java/com/atakmap/map/layer/raster/mosaic/MosaicDatabase2.java
package com.atakmap.map.layer.raster.mosaic;

import java.io.File;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

import com.atakmap.coremap.maps.coords.GeoPoint;
import com.atakmap.database.RowIterator;
import com.atakmap.map.layer.feature.geometry.Envelope;
import com.atakmap.map.layer.feature.geometry.Geometry;
import com.atakmap.map.layer.raster.ImageInfo;

public interface MosaicDatabase2 {

    public String getType();
    public void open(File f);
    public void close();
    public Coverage getCoverage();
    public void getCoverages(Map<String, Coverage> coverages);
    public Coverage getCoverage(String type);
    public Cursor query(QueryParameters params);

    /**************************************************************************/

    public static class QueryParameters {
        public static enum GsdCompare {
            MinimumGsd,
            MaximumGsd,
        }
        public static enum Order {
            MinGsdAsc,
            MinGsdDesc,
            MaxGsdAsc,
            MaxGsdDesc,
        }

        public String path;
        public Geometry spatialFilter;
        public double minGsd;
        public double maxGsd;
        public Set<String> types;
        public int srid;
        public Boolean precisionImagery;
        public GsdCompare minGsdCompare = GsdCompare.MaximumGsd;
        public GsdCompare maxGsdCompare = GsdCompare.MaximumGsd;
        public Order order = Order.MaxGsdDesc;

        public QueryParameters() {
            this.path = null;
            this.spatialFilter = null;
            this.minGsd = Double.NaN;
            this.maxGsd = Double.NaN;
            this.types = null;
            this.srid = -1;
            this.precisionImagery = null;
        }
        
        public QueryParameters(QueryParameters other) {
            this.path = other.path;
            // XXX - needs to be cloned
            this.spatialFilter = other.spatialFilter;
            this.minGsd = other.minGsd;
            this.maxGsd = other.maxGsd;
            if(other.types != null)
                this.types = new HashSet<String>(other.types);
            else
                this.types = null;
            this.srid = other.srid;
            this.precisionImagery = null;
            this.minGsdCompare = other.minGsdCompare;
            this.maxGsdCompare = other.maxGsdCompare;
        }
    }

    public static interface Cursor extends RowIterator {
        public final static Cursor EMPTY = new Cursor() {

            @Override
            public boolean moveToNext() { return false; }
            @Override
            public void close() {}
            @Override
            public boolean isClosed() { return false; }
            @Override
            public GeoPoint getUpperLeft() { return null; }
            @Override
            public GeoPoint getUpperRight() { return null; }
            @Override
            public GeoPoint getLowerRight() { return null; }
            @Override
            public GeoPoint getLowerLeft() { return null; }
            @Override
            public double getMinLat() { return 0; }
            @Override
            public double getMinLon() { return 0; }
            @Override
            public double getMaxLat() { return 0; }
            @Override
            public double getMaxLon() { return 0; }
            @Override
            public String getPath() { return null; }
            @Override
            public String getType() { return null; }
            @Override
            public double getMinGSD() { return 0; }
            @Override
            public double getMaxGSD() { return 0; }
            @Override
            public int getWidth() { return 0; }
            @Override
            public int getHeight() { return 0; }
            @Override
            public int getId() { return 0; }
            @Override
            public int getSrid() { return -1; }
            @Override
            public boolean isPrecisionImagery() { return false; }
            @Override
            public Frame asFrame() { return null; }
        };

        public abstract GeoPoint getUpperLeft();
        public abstract GeoPoint getUpperRight();
        public abstract GeoPoint getLowerRight();
        public abstract GeoPoint getLowerLeft();
        public abstract double getMinLat();
        public abstract double getMinLon();
        public abstract double getMaxLat();
        public abstract double getMaxLon();
        public abstract String getPath();
        public abstract String getType();
        public abstract double getMinGSD();
        public abstract double getMaxGSD();
        public abstract int getWidth();
        public abstract int getHeight();
        public abstract int getId();
        public int getSrid();
        public boolean isPrecisionImagery();
        public MosaicDatabase2.Frame asFrame();
    }
    
    public final static class Frame extends ImageInfo {
        public final int id;
        public final double minLat;
        public final double minLon;
        public final double maxLat;
        public final double maxLon;
        public final double minGsd;

        public Frame(int id, String path, String type, boolean precisionImagery, double minLat, double minLon, double maxLat,
                double maxLon, GeoPoint upperLeft, GeoPoint upperRight, GeoPoint lowerRight,
                GeoPoint lowerLeft, double minGsd, double maxGsd, int width, int height, int srid) {
            super(path, type, precisionImagery, upperLeft, upperRight, lowerRight, lowerLeft, maxGsd, width, height, srid);
            
            this.id = id;
            this.minLat = minLat;
            this.minLon = minLon;
            this.maxLat = maxLat;
            this.maxLon = maxLon;
            this.minGsd = minGsd;
        }
        
        public Frame(Cursor row) {
            this(row.getId(),
                 row.getPath(),
                 row.getType(),
                 row.isPrecisionImagery(),
                 row.getMinLat(),
                 row.getMinLon(),
                 row.getMaxLat(),
                 row.getMaxLon(),
                 row.getUpperLeft(),
                 row.getUpperRight(),
                 row.getLowerRight(),
                 row.getLowerLeft(),
                 row.getMinGSD(),
                 row.getMaxGSD(),
                 row.getWidth(),
                 row.getHeight(),
                 row.getSrid());
        }
    }

    public final static class Coverage {
        public final Geometry geometry;
        public final double minGSD;
        public final double maxGSD;

        public Coverage(Geometry geometry, double minGSD, double maxGSD) {
            this.geometry = geometry;
            this.minGSD = minGSD;
            this.maxGSD = maxGSD;
        }

        public String toString() {
            final Envelope mbb = this.geometry.getEnvelope();
            return "Coverage {minLat=" + mbb.minY + ",minLon=" + mbb.minX + ",maxLat=" + mbb.maxY
                    + ",maxLon=" + mbb.maxX + ",minGSD=" + minGSD + ",maxGSD=" + maxGSD + "}";
        }
    }
}

File: takengine/takengine/src/main/cpp/raster/mosaic/MosaicDatabase2.cpp
#include "raster/mosaic/MosaicDatabase2.h"

#include <cmath>

#include "port/Collections.h"
#include "port/STLSetAdapter.h"
#include "port/String.h"
#include "util/Memory.h"

using namespace TAK::Engine;
using namespace TAK::Engine::Core;
using namespace TAK::Engine::Feature;
using namespace TAK::Engine::Raster::Mosaic;
using namespace TAK::Engine::Util;

namespace
{
    double min(double a, double b, double c, double d)
    {
        double     v = a;
        if (b < v) v = b;
        if (c < v) v = c;
        if (d < v) v = d;
        return v;
    }

    double max(double a, double b, double c, double d)
    {
        double     v = a;
        if (b > v) v = b;
        if (c > v) v = c;
        if (d > v) v = d;
        return v;
    }
}

MosaicDatabase2::~MosaicDatabase2() NOTHROWS
{}

/**************************************************************************/
// MosaicDatabase2::QueryParameters

MosaicDatabase2::QueryParameters::QueryParameters() NOTHROWS :
    path(nullptr),
    spatialFilter(nullptr, nullptr),
    minGsd(NAN),
    maxGsd(NAN),
    types(nullptr, nullptr),
    srid(-1),
    imagery(QueryParameters::AllImagery),
    minGsdCompare(MaximumGsd),
    maxGsdCompare(MaximumGsd),
    order(MaxGsdDesc)
{}
        
MosaicDatabase2::QueryParameters::QueryParameters(const QueryParameters &other) NOTHROWS :
    path(other.path),
    spatialFilter(nullptr, nullptr),
    minGsd(other.minGsd),
    maxGsd(other.maxGsd),
    types(nullptr, nullptr),
    srid(other.srid),
    imagery(other.imagery),
    minGsdCompare(other.minGsdCompare),
    maxGsdCompare(other.maxGsdCompare),
    order(other.order)
{
    if (other.spatialFilter.get()) {
        TAKErr code(TE_Ok);
        code = Geometry_clone(spatialFilter, *other.spatialFilter);
        if (code != TE_Ok) {
            // XXX - 
        }
    }

    if (other.types.get()) {
        types = TAK_UNIQUE_PTR(Port::Set<Port::String>)(new Port::STLSetAdapter<Port::String, Port::StringLess>(), Util::Memory_deleter_const<Port::Set<Port::String>, Port::STLSetAdapter<Port::String, Port::StringLess>>);
        Port::Collections_addAll(*types, *other.types);
    }
}

MosaicDatabase2::QueryParameters::~QueryParameters() NOTHROWS
{}

MosaicDatabase2::Cursor::~Cursor() NOTHROWS
{}


MosaicDatabase2::Frame::Frame(const int id_,
                              const char *path_,
                              const char *type_,
                              const bool precisionImagery_,
                              const GeoPoint2 &upperLeft_,
                              const GeoPoint2 &upperRight_,
                              const GeoPoint2 &lowerRight_,
                              const GeoPoint2 &lowerLeft_,
                              const double minGsd_,
                              const double maxGsd_,
                              const int width_,
                              const int height_,
                              const int srid_) NOTHROWS :
    Raster::ImageInfo(path_,
                      type_,
                      precisionImagery_,
                      upperLeft_,
                      upperRight_,
                      lowerRight_,
                      lowerLeft_,
                      maxGsd_,
                      width_,
                      height_,
                      srid_),
    id(id_),
    minLat(min(upperLeft_.latitude, upperRight_.latitude, lowerRight_.latitude, lowerLeft_.latitude)),
    minLon(min(upperLeft_.longitude, upperRight_.longitude, lowerRight_.longitude, lowerLeft_.longitude)),
    maxLat(max(upperLeft_.latitude, upperRight_.latitude, lowerRight_.latitude, lowerLeft_.latitude)),
    maxLon(max(upperLeft_.longitude, upperRight_.longitude, lowerRight_.longitude, lowerLeft_.longitude)),
    minGsd(minGsd_)
{}
    
MosaicDatabase2::Frame::Frame(const int id_,
                              const char *path_,
                              const char *type_,
                              const bool precisionImagery_,
                              const double minLat_,
                              const double minLon_,
                              const double maxLat_,
                              const double maxLon_,
                              const GeoPoint2 &upperLeft_,
                              const GeoPoint2 &upperRight_,
                              const GeoPoint2 &lowerRight_,
                              const GeoPoint2 &lowerLeft_,
                              const double minGsd_,
                              const double maxGsd_,
                              const int width_,
                              const int height_,
                              const int srid_) NOTHROWS :
    Raster::ImageInfo(path_,
                      type_,
                      precisionImagery_,
                      upperLeft_,
                      upperRight_,
                      lowerRight_,
                      lowerLeft_,
                      maxGsd_,
                      width_,
                      height_,
                      srid_),
    id(id_),
    minLat(minLat_),
    minLon(minLon_),
    maxLat(maxLat_),
    maxLon(maxLon_),
    minGsd(minGsd_)
{}

Util::TAKErr MosaicDatabase2::Frame::createFrame(MosaicDatabase2::FramePtr_const &frame, MosaicDatabase2::Cursor &row) NOTHROWS
{
    TAKErr code(TE_Ok);

    int id;
    code = row.getId(&id);
    TE_CHECKRETURN_CODE(code);

    const char *path;
    code = row.getPath(&path);
    TE_CHECKRETURN_CODE(code);

    const char *type;
    code = row.getType(&type);
    TE_CHECKRETURN_CODE(code);

    bool precisionImagery;
    code = row.isPrecisionImagery(&precisionImagery);
    TE_CHECKRETURN_CODE(code);

    double minLat;
    code = row.getMinLat(&minLat);
    TE_CHECKRETURN_CODE(code);

    double minLon;
    code = row.getMinLon(&minLon);
    TE_CHECKRETURN_CODE(code);

    double maxLat;
    code = row.getMaxLat(&maxLat);
    TE_CHECKRETURN_CODE(code);

    double maxLon;
    code = row.getMaxLon(&maxLon);
    TE_CHECKRETURN_CODE(code);

    GeoPoint2 upperLeft;
    code = row.getUpperLeft(&upperLeft);
    TE_CHECKRETURN_CODE(code);

    GeoPoint2 upperRight;
    code = row.getUpperRight(&upperRight);
    TE_CHECKRETURN_CODE(code);

    GeoPoint2 lowerRight;
    code = row.getLowerRight(&lowerRight);
    TE_CHECKRETURN_CODE(code);

    GeoPoint2 lowerLeft;
    code = row.getLowerLeft(&lowerLeft);
    TE_CHECKRETURN_CODE(code);

    double minGsd;
    code = row.getMinGSD(&minGsd);
    TE_CHECKRETURN_CODE(code);

    double maxGsd;
    code = row.getMaxGSD(&maxGsd);
    TE_CHECKRETURN_CODE(code);

    int width;
    code = row.getWidth(&width);
    TE_CHECKRETURN_CODE(code);

    int height;
    code = row.getHeight(&height);
    TE_CHECKRETURN_CODE(code);

    int srid;
    code = row.getSrid(&srid);
    TE_CHECKRETURN_CODE(code);

    frame = FramePtr_const(new Frame(id,
                                     path,
                                     type,
                                     precisionImagery,
                                     minLat,
                                     minLon,
                                     maxLat,
                                     maxLon,
                                     upperLeft,
                                     upperRight,
                                     lowerRight,
                                     lowerLeft,
                                     minGsd,
                                     maxGsd,
                                     width,
                                     height,
                                     srid),
                           Memory_deleter_const<Frame>);

    return code;
}


MosaicDatabase2::Coverage::Coverage(Geometry2Ptr_const &&geometry_, const double minGSD_, const double maxGSD_) NOTHROWS :
    geometry(std::move(geometry_)),
    minGSD(minGSD_),
    maxGSD(maxGSD_)
{}
